be/src/exec/pipeline/scan/morsel.cpp

```diff
@@ -229,5 +238,315 @@ Status PhysicalSplitMorselQueue::_init_segment() {
+        }
+    } else {
+        Slice short_key = *_next_lower_block_iter;
+        return std::make_unique<vectorized::ShortKeyOption>(_short_key_schema, short_key, false);
+    }
+}
+
+bool LogicalSplitMorselQueue::_cur_tablet_finished() const {
+    return _range_idx >= _block_ranges_per_seek_range.size();
+}
+
+Rowset* LogicalSplitMorselQueue::_find_largest_rowset(const std::vector<RowsetSharedPtr>& rowsets) {
+    if (rowsets.empty()) {
+        return nullptr;
+    }
+
+    Rowset* largest_rowset = rowsets[0].get();
+    for (int i = 1; i < rowsets.size(); ++i) {
+        if (largest_rowset->num_rows() < rowsets[i]->num_rows()) {
+            largest_rowset = rowsets[i].get();
+        }
+    }
+
+    return largest_rowset;
+}
+
+SegmentSharedPtr LogicalSplitMorselQueue::_find_largest_segment(Rowset* rowset) const {
+    auto* beta_rowset = down_cast<BetaRowset*>(rowset);
+    const auto& segments = beta_rowset->segments();
+    if (segments.empty()) {
+        return nullptr;
+    }
+
+    SegmentSharedPtr largest_segment = segments[0];
+    for (int i = 1; i < segments.size(); ++i) {
+        if (largest_segment->num_rows() < segments[i]->num_rows()) {
+            largest_segment = segments[i];
+        }
+    }
+
+    return largest_segment;
+}
+
+StatusOr<SegmentGroupPtr> LogicalSplitMorselQueue::_create_segment_group(Rowset* rowset) {
+    std::vector<SegmentSharedPtr> segments;
+    if (rowset->rowset_meta()->is_segments_overlapping()) {
+        segments.emplace_back(_find_largest_segment(rowset));
+    } else {
+        auto* beta_rowset = down_cast<BetaRowset*>(rowset);
+        segments = beta_rowset->segments();
+    }
+
+    for (const auto& segment : segments) {
+        RETURN_IF_ERROR(segment->load_index(StorageEngine::instance()->tablet_meta_mem_tracker()));
+    }
+
+    return std::make_unique<SegmentGroup>(std::move(segments));
+}
+
+bool LogicalSplitMorselQueue::_next_tablet() {
+    if (!_has_init_any_tablet) {
+        _has_init_any_tablet = true;
+    } else {
+        ++_tablet_idx;
+    }
+
+    return _tablet_idx < _tablets.size();
+}
+
+Status LogicalSplitMorselQueue::_init_tablet() {
+    _largest_rowset = nullptr;
+    _segment_group = nullptr;
+    _short_key_schema = nullptr;
+    _block_ranges_per_seek_range.clear();
+    _num_rest_blocks_per_seek_range.clear();
+    _range_idx = 0;
+
+    if (_tablet_idx == 0) {
+        // All the tablets have the same schema, so parse seek range with the first table schema.
+        RETURN_IF_ERROR(vectorized::TabletReader::parse_seek_range(_tablets[_tablet_idx], _range_start_op,
+                                                                   _range_end_op, _range_start_key, _range_end_key,
+                                                                   &_tablet_seek_ranges, &_mempool));
+    }
+
+    _largest_rowset = _find_largest_rowset(_tablet_rowsets[_tablet_idx]);
+    if (_largest_rowset == nullptr || _largest_rowset->num_rows() == 0) {
```

Comment:
Is it safe to ignore the initialization of the following fields when no rowset exist?

Reply:
Yes, if there isn't rowset, `_segment_group` is nullptr, and will go forward next rowset in `try_get`.